orbstation/code/customisation/preferences/languages.dm

```diff
@@ -0,0 +1,172 @@
+/datum/preferences/proc/get_adjusted_language_holder()
+	var/datum/species/species = read_preference(/datum/preference/choiced/species)
+	species = new species()
+	var/datum/language_holder/language_holder = new species.species_language_holder()
+
+	return language_holder
+
+/datum/asset/spritesheet/languages
+	name = "languages"
+	early = TRUE
+	cross_round_cachable = TRUE
+
+/datum/asset/spritesheet/languages/create_spritesheets()
+	var/list/to_insert = list()
+
+	if(!GLOB.all_languages.len)
+		for(var/iterated_language in subtypesof(/datum/language))
+			var/datum/language/language = iterated_language
+			if(!initial(language.key))
+				continue
+
+			GLOB.all_languages += language
+
+			var/datum/language/instance = new language
+
+			GLOB.language_datum_instances[language] = instance
+
+	for (var/language_name in GLOB.all_languages)
+		var/datum/language/language = GLOB.language_datum_instances[language_name]
+		var/icon/language_icon = icon(language.icon, icon_state = language.icon_state)
+		to_insert[sanitize_css_class_name(language.name)] = language_icon
+
+	for (var/spritesheet_key in to_insert)
+		Insert(spritesheet_key, to_insert[spritesheet_key])
+
+/// Middleware to handle languages
+/datum/preference_middleware/languages
+	/// A associative list of language names to their typepath
+	var/static/list/name_to_language = list()
+	action_delegations = list(
+		"give_language" = PROC_REF(give_language),
+		"remove_language" = PROC_REF(remove_language),
+	)
+
+/datum/preference_middleware/languages/apply_to_human(mob/living/carbon/human/target, datum/preferences/preferences)
+	var/datum/language_holder/language_holder = target.get_language_holder()
+	language_holder.remove_all_languages()
+	for(var/lang_path in preferences.languages)
+		language_holder.grant_language(lang_path)
+
+/datum/preference_middleware/languages/get_ui_assets()
+	return list(
+		get_asset_datum(/datum/asset/spritesheet/languages),
+	)
+
+/datum/preference_middleware/languages/post_set_preference(mob/user, preference, value)
+	if(preference != "species")
+		return
+	preferences.languages = list()
+	var/species_type = preferences.read_preference(/datum/preference/choiced/species)
+	var/datum/species/species = new species_type()
+	var/datum/language_holder/lang_holder = new species.species_language_holder()
+	for(var/language in preferences.get_adjusted_language_holder())
+		preferences.languages[language] = LANGUAGE_SPOKEN
+	qdel(lang_holder)
+	qdel(species)
+
+	for(var/language in lang_holder.spoken_languages)
+		preferences.languages[language] = LANGUAGE_SPOKEN
+
+	qdel(lang_holder)
+	qdel(species)
+
+	return ..()
+
+/datum/preference_middleware/languages/get_ui_data(mob/user)
+	if(length(name_to_language) != length(GLOB.all_languages))
+		initialize_name_to_language()
+
+	var/list/data = list()
+
+	var/max_languages = preferences.all_quirks.Find(QUIRK_LINGUIST) ? MAX_LANGUAGES_LINGUIST : MAX_LANGUAGES_NORMAL
```

Comment:
Can you search for the `QUIRK_LINGUIST` trait instead or is it not applied yet at this point?

Reply:
Actually, this was supposed to be searching for the _name_ of the Linguist trait, which I guess Skyrat put in a define in case it ever changed (but then they also made it the actual quirk trait?? I don't know). I'm just putting the quirk name here, and if that's bad I can do it as a define again.